scylla/src/statement/bound.rs

```diff
@@ -0,0 +1,59 @@
+use scylla_cql::serialize::{
+    row::{SerializeRow, SerializedValues},
+    SerializationError,
+};
+
+use crate::routing::Token;
+
+use super::prepared::{
+    PartitionKey, PartitionKeyError, PartitionKeyExtractionError, PreparedStatement,
+};
+
+/// Represents a statement that already had all its values bound
+#[derive(Debug, Clone)]
+pub struct BoundStatement {
+    pub(crate) prepared: PreparedStatement,
+    pub(crate) values: SerializedValues,
+}
+
+impl BoundStatement {
+    pub(crate) fn new(
+        prepared: PreparedStatement,
+        values: &impl SerializeRow,
+    ) -> Result<BoundStatement, SerializationError> {
+        let values = prepared.serialize_values(values)?;
+        Ok(Self { prepared, values })
+    }
+
+    /// Determines which values constitute the partition key and puts them in order.
+    ///
+    /// This is a preparation step necessary for calculating token based on a prepared statement.
+    pub(crate) fn pk(&self) -> Result<PartitionKey<'_>, PartitionKeyExtractionError> {
+        PartitionKey::new(self.prepared.get_prepared_metadata(), &self.values)
+    }
+
+    pub(crate) fn pk_and_token(
+        &self,
+    ) -> Result<Option<(PartitionKey<'_>, Token)>, PartitionKeyError> {
+        if !self.prepared.is_token_aware() {
+            return Ok(None);
+        }
+
+        let partition_key = self.pk()?;
+        let token = partition_key.calculate_token(self.prepared.get_partitioner_name())?;
+        Ok(Some((partition_key, token)))
+    }
+
+    /// Calculates the token for the prepared statement and its bound values
+    ///
+    /// Returns the token that would be computed for executing the provided prepared statement with
+    /// the provided values.
+    pub fn token(&self) -> Result<Option<Token>, PartitionKeyError> {
+        self.pk_and_token().map(|p| p.map(|(_, t)| t))
+    }
```

Comment:
:wrench: The previous method names (e.g., `extract_partition_key_and_calculate_token()`), were verbose but informative. Their names warned that they required computations; the new names suggest getter-like cheap operation.

Reply:
Fair enough, and I am okay changing it back but if I could push back just slightly, the previous one was more expensive because it also serialized the values as part of that calculation. For this method the values have already been serialized so it is a cheaper calculation.